consensus/bor/heimdall/client.go

```diff
@@ -126,58 +131,69 @@ func (h *HeimdallClient) FetchLatestCheckpoint() (*checkpoint.Checkpoint, error)
 }
 
 // FetchWithRetry returns data from heimdall with retry
-func FetchWithRetry[T any](client http.Client, url *url.URL, closeCh chan struct{}) (*T, error) {
-	// attempt counter
-	attempt := 1
-	result := new(T)
-
-	ctx, cancel := context.WithTimeout(context.Background(), apiHeimdallTimeout)
-
+func FetchWithRetry[T any](ctx context.Context, client http.Client, url *url.URL, closeCh chan struct{}) (*T, error) {
 	// request data once
-	body, err := internalFetch(ctx, client, url)
-
-	cancel()
-
-	if err == nil && body != nil {
-		err = json.Unmarshal(body, result)
-		if err != nil {
-			return nil, err
-		}
-
+	result, err := Fetch[T](ctx, client, url)
+	if err == nil {
 		return result, nil
 	}
 
+	// ignore or log the error
+
 	// create a new ticker for retrying the request
-	ticker := time.NewTicker(5 * time.Second)
+	ticker := time.NewTicker(retryCall)
 	defer ticker.Stop()
 
+	// attempt counter
+	attempt := 1
+
 	for {
 		log.Info("Retrying again in 5 seconds to fetch data from Heimdall", "path", url.Path, "attempt", attempt)
+
 		attempt++
+
 		select {
+		case <-ctx.Done():
+			log.Debug("Shutdown detected, terminating request")
+
+			return nil, ctx.Err()
 		case <-closeCh:
 			log.Debug("Shutdown detected, terminating request")
 
-			return nil, errShutdownDetected
+			return nil, ErrShutdownDetected
 		case <-ticker.C:
-			ctx, cancel = context.WithTimeout(context.Background(), apiHeimdallTimeout)
-
-			body, err = internalFetch(ctx, client, url)
-
-			cancel()
-
-			if err == nil && body != nil {
-				err = json.Unmarshal(body, result)
-				if err != nil {
-					return nil, err
-				}
+			result, err = Fetch[T](ctx, client, url)
 
+			// ignore or log the error
+			if err == nil {
 				return result, nil
 			}
 		}
 	}
 }
 
+// Fetch returns data from heimdall
+func Fetch[T any](ctx context.Context, client http.Client, url *url.URL) (*T, error) {
+	result := new(T)
+
+	body, err := internalFetchWithTimeout(ctx, client, url)
+
+	if body == nil {
+		return nil, ErrNoResponse
+	}
+
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Should we replace the above code blocks with each other? Isn't there a possibility of having a nil body with some error and if yes, the we won't be able to send the error to the parent function.